Sources/_StringProcessing/RegexDSL/Core.swift

```diff
@@ -13,39 +11,47 @@ public protocol RegexProtocol {
   var regex: Regex<Capture> { get }
 }
 
-public protocol EmptyProtocol {}
-public struct Empty: EmptyProtocol {}
-extension Array: EmptyProtocol where Element: EmptyProtocol {}
-extension Optional: EmptyProtocol where Wrapped: EmptyProtocol {}
-
 /// A regular expression.
 public struct Regex<Capture>: RegexProtocol {
   /// A program representation that caches any lowered representation for
   /// execution.
   internal class Program {
-    /// The underlying AST.
-    let ast: AST
+    /// The underlying pattern tree
+    let tree: PatternTree
+
+    var regexAST: AST? { tree.regexAST }
+
     /// The legacy `RECode` for execution with a legacy VM.
     lazy private(set) var legacyLoweredProgram: RECode = {
       do {
+        guard let ast = regexAST else {
+          fatalError("FIXME")
+        }
         return try compile(ast)
       } catch {
         fatalError("Regex engine internal error: \(String(describing: error))")
       }
     }()
     /// The program for execution with the matching engine.
-    lazy private(set) var loweredProgram = Compiler(ast: ast).emit()
+    lazy private(set) var loweredProgram = Compiler(
+      ast: regexAST! // FIXME
+    ).emit()
 
-    init(ast: AST) {
-      self.ast = ast
+    init(_ tree: PatternTree) {
+      self.tree = tree
+    }
+    convenience init(_ ast: AST) {
+      self.init(.regex(ast))
     }
   }
 
   let program: Program
-  var ast: AST { program.ast }
 
-  init(ast: AST) {
-    self.program = Program(ast: ast)
+  var tree: PatternTree { program.tree }
+  var regexAST: AST? { program.regexAST }
```

Comment:
If we kept its name `ast` and made it non-optional (it okay to fail for now since it’s to be overhauled again), most of the changes to the gyb’d contents will not be necessary.

Reply:
But we'd still need to change them when we stop using the AST, right?